---
title: "Fig5"
author: "Christine Gao"
date: "5/16/2023"
output: 
  html_document:
    code_folding: hide
---
```{r, include=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(tximport)
library(DESeq2)
library(sva)
library(biomaRt)
library(ggplot2)
library(ggrepel)
library(tximeta)
library(SummarizedExperiment)
library(edgeR)
library(limma)
library(statmod)
library(tidyverse)
library(tripack)
library(EnsDb.Mmusculus.v79)
library(BSgenome.Mmusculus.UCSC.mm10)
library(dplyr)
library(scales)
```

Function for making volcano plots FOR PAPER FIGURES.
```{r}
makeVolcanoPlot.figure <- 
  function(DEgenes_df, lfc_cutoff, fdr, 
           gene_ids1, gene_ids2, 
           circ_ids, circ_names, ylim, 
           color1, color2, title){
  tab = data.frame(logFC = DEgenes_df$log2FoldChange, negLogPval = -log10(DEgenes_df$pvalue))
  rownames(tab) <- rownames(DEgenes_df)
  par(mar=c(1.4,1.3,1,0.5), mgp=c(0.5,0,0))
  plot(tab[-which(rownames(tab) %in% c(gene_ids1, gene_ids2)), ], pch = 19, cex = 0.25, 
       xlab='',
       ylab='',
       axes = FALSE,
       col = "gray60", bty = 'l', 
       xlim = c(-4, 4), 
       ylim = c(0, ylim))
  points(tab[gene_ids1, ], pch = 19, cex = 0.3, col = color1)
  points(tab[gene_ids2, ], pch = 19, cex = 0.3, col = color2)
  points(tab[circ_ids, ], pch = 1, cex = 1, col= 'black')
  
  axis(1, seq(-4,6, by=2), tck=-0.02, cex.axis=0.7)
  axis(2, seq(0,10, by=2), tck=-0.02, cex.axis=0.7)
  mtext(expression(log[2]~fold~change), side=1, line=0.6, cex=0.7)
  mtext(expression(-log[10]~pvalue), side=2, line=0.6, cex=0.7)
  mtext(title, side=3, cex=0.7)
  
  yJitter <- tab[circ_ids, ]$negLogPval
  yJitter[2] <- yJitter[2]-0.5
  yJitter[4] <- yJitter[4]+0.2
  # for (i in yJitter){
  #   yJitter[i]<- (yJitter[i]+runif(1,-.5,.5))
  #   }
  text(tab[circ_ids, ]$logFC-0.4, yJitter, adj=1, cex = 0.6, col='black',font=3,
       labels=circ_names)
  # text(tab[circ_ids, ]$logFC+0.2, tab[circ_ids, ]$negLogPval, adj=0, cex = 0.6, col='black',font=3,
  #      labels=circ_names)
 
  abline(h=-log10(fdr), col='red', lty=2, lwd=1)
}
```

Function for retrieving gene names, given a vector of Ensembl IDs. Returns a dataframe containing both Ensembl IDs and corresponding genes. <i>**Leaves out Ensembl IDs for which a gene name could not be found.</i>
```{r}
get.gene <- function(x){
  ensembl102.mmusculus <- useEnsembl(biomart='genes', dataset='mmusculus_gene_ensembl', version=102)
  mgi <- getBM(attributes=c('ensembl_gene_id','mgi_symbol'),
               filters='ensembl_gene_id',
               values=x,
               mart=ensembl102.mmusculus)
  genes <- data.frame(ensembl_gene_id=x)
  genes <- merge(genes, mgi, by='ensembl_gene_id')
  return(genes)
}
```

```{r}
get.ensembl <- function(x) {
  edb_mouse <- EnsDb.Mmusculus.v79
  proms_mouse <- promoters(edb_mouse, 
                           filter = TxBiotypeFilter("protein_coding"), 
                           upstream = 2000, 
                           downstream = 2000, 
                           columns = c("gene_name", "tx_id", "gene_id"))
  genome(seqinfo(proms_mouse)) <- "mm10"
  seqlevelsStyle(proms_mouse) <- "ucsc"
  proms_mouse <- proms_mouse[which(seqnames(proms_mouse) %in% seqnames(Mmusculus)[1:21])]
  proms_mouse$gene_name <- toupper(proms_mouse$gene_name)

  x <- toupper(x)
  x <- unique(x)
  ensembl.ids <- as.data.frame(
    unique(proms_mouse[which(proms_mouse$gene_name %in% x), c('gene_name','gene_id')]),
    row.names=NULL)
  ensembl.ids <- ensembl.ids[, c('gene_name','gene_id')] %>% distinct()
  return(ensembl.ids)
}
```

<b>Fig 5D</b>
```{r}
load(file='BigStuff/dds.coll.filtered_GSKJ4.rda')

plot <- plotPCA(vst(dds.coll.filtered), intgroup=c('genotype','tx'), returnData=TRUE)

# setEPS()
# postscript('plots_final/pca_GSKJ4.eps', width=2.15, height=1.57)
plot$group <- factor(plot$group, levels=c('WT:DMSO','WT:GSKJ4','Mut:DMSO','Mut:GSKJ4'))
group.colors <- c('#0F80FF','#0F80FF','#FB0207','#FB0207')
ggplot(data = plot, aes(x = PC1, y = PC2, color=group, shape=group)) + 
  geom_point(size = 2) + 
  scale_shape_manual(values=c(16,1,17,2)) +
  scale_color_manual(values=group.colors) + 
  xlab('PC1: 77% variance') +
  ylab('PC2: 6% variance') +
  theme_bw(base_size=8) +
  theme(axis.text=element_text(size=8, color='black'),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_rect(color='black',
                                    fill=NA,
                                    size=1),
        legend.position = 'none')
# dev.off()
```

<b>Fig 5E</b>
```{r}
load(file='BigStuff/dds_final.rda')
res <- results(dds)

load(file='BigStuff/dds_final_GSKJ4.rda')
res.MutDMSO.WtDMSO <- results(dds, contrast = list(c('genotype_Mut_vs_WT'))) # Mut_DMSO vs WT_DMSO

sig.MutDMSO.WtDMSO <- rownames(unique(res.MutDMSO.WtDMSO[which(res.MutDMSO.WtDMSO$padj<0.1),]))

# setEPS()
# postscript(file='plots_final/density_cond-pvalue.eps', width=1.9, height=1.57)
par(mar=c(1.35,1.1,0,0), mgp=c(0.5,0.075,0))
plot(density(res[which(rownames(res) %in% sig.MutDMSO.WtDMSO),]$pvalue, from=0.01,to=0.97, bw=0.025), 
     asp=NA, lwd=2, col='orange', main='', bty='n', yaxt='n', xaxt='n', ylab='', xlab='')
axis(1, at=c(0,1), tck=-0.02, cex.axis=0.7)
axis(2, at=c(.7,2), tck=-0.02, cex.axis=0.7, pos=-0.02)
mtext('p-values at D14', side=1, line=0.3, cex=0.7)
mtext('Density', side=2, line=0.3, cex=0.7)

lines(density(res[which(!(rownames(res) %in% sig.MutDMSO.WtDMSO)),]$pvalue, 
              from=0.01,to=0.97, bw=0.025),
      lwd=2, col='black')

legend('topright', legend=c('p-adj > 0.1, D21', 'p-adj < 0.1, D21'),
       col=c("black", "orange"), lty=c(1,1), lwd=2, cex=0.7,
       box.lty=0)

# dev.off()
```

<b>Fig 5F</b>
```{r}
load(file='mgi_ost.condense.ensembl.rda')
load(file='mgi_condense.ensembl.rda')

fdr.MutDMSO.WtDMSO <- res.MutDMSO.WtDMSO[which(res.MutDMSO.WtDMSO$padj>0.1 
                                  & is.na(res.MutDMSO.WtDMSO$padj)==FALSE),]
fdr.MutDMSO.WtDMSO <- fdr.MutDMSO.WtDMSO[
  which(fdr.MutDMSO.WtDMSO$padj==min(fdr.MutDMSO.WtDMSO$padj)),]$pvalue[1]

circ_ids <- get.ensembl(c('Satb2','Dlx3','Pth1r','Runx2'))
circ_names <- get.gene(circ_ids$gene_id)

# setEPS()
# postscript('plots_final/volcano_ost-bmp_MutDMSO_WtDMSO.eps', width=1.96, height=1.96)
makeVolcanoPlot.figure(res.MutDMSO.WtDMSO,
                "none", fdr.MutDMSO.WtDMSO,
                mgi_ost.condense.ensembl$gene_id,
                mgi_condense.ensembl,
                circ_ids$gene_id, circ_names$mgi_symbol,
                ylim=10, "deepskyblue", "magenta", 'R684C/+ DMSO vs +/+ DMSO')
# dev.off()
```

<b>Fig 5G</b>
```{r}
res.MutGSKJ4.MutDMSO <- results(dds, contrast = list(c('tx_GSKJ4_vs_DMSO',
                                                       'genotypeMut.txGSKJ4'))) # Mut_GSKJ4 vs Mut_DMSO
fdr.MutGSKJ4.MutDMSO <- res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$padj>0.1 
                                  & is.na(res.MutGSKJ4.MutDMSO$padj)==FALSE),]
fdr.MutGSKJ4.MutDMSO <- fdr.MutGSKJ4.MutDMSO[
  which(fdr.MutGSKJ4.MutDMSO$padj==min(fdr.MutGSKJ4.MutDMSO$padj)),]$pvalue[1]

circ_ids <- get.ensembl(c('Satb2','Dlx3','Pth1r','Runx2'))
circ_names <- get.gene(circ_ids$gene_id)

# setEPS()
# postscript('plots_final/volcano_ost-bmp_MutGSKJ4_MutDMSO.eps', width=1.96, height=1.96)
makeVolcanoPlot.figure(res.MutGSKJ4.MutDMSO,
                "none", fdr.MutGSKJ4.MutDMSO,
                mgi_ost.condense.ensembl$gene_id,
                mgi_condense.ensembl,
                circ_ids$gene_id, circ_names$mgi_symbol,
                ylim=10, "deepskyblue", "magenta", 'R684C/+ GSKJ4 vs R684C/+ DMSO')
# dev.off()
```

<b>Fig 5H</b>
```{r}
log2FC.MutDMSO.WtDMSO <- data.frame(
  log2FC.MutDMSO.WtDMSO=res.MutDMSO.WtDMSO[which(
    res.MutDMSO.WtDMSO$padj<0.1),]$log2FoldChange,
  ensembl_id=rownames(res.MutDMSO.WtDMSO[which(res.MutDMSO.WtDMSO$padj<0.1),]))

log2FC.MutGSKJ4.MutDMSO <- log2FC.MutGSKJ4.MutDMSO <- data.frame(
  log2FC.MutGSKJ4.MutDMSO=res.MutGSKJ4.MutDMSO[which(
    res.MutGSKJ4.MutDMSO$padj<0.1),]$log2FoldChange,
  ensembl_id=rownames(res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$padj<0.1),]))

df.compare <- merge(log2FC.MutDMSO.WtDMSO, log2FC.MutGSKJ4.MutDMSO, by='ensembl_id')

setEPS()
postscript(file='plots_final/log2FC-reversal.eps', width=2.26, height=1.57)
par(mar=c(1.5,2,0.3,0.3), mgp=c(0.5,0,0))
plot(df.compare$log2FC.MutDMSO.WtDMSO, df.compare$log2FC.MutGSKJ4.MutDMSO,
     pch=16, cex=0.5, col='gray60',
     xlim=c(-3.5,3), ylim=c(-3,3),
     xlab = '', ylab='', axes=FALSE, 
     main='')
points(df.compare$log2FC.MutDMSO.WtDMSO[which(df.compare$ensembl %in%
                                                mgi_ost.condense.ensembl$gene_id)],
       df.compare$log2FC.MutGSKJ4.MutDMSO[which(df.compare$ensembl %in%
                                                  mgi_ost.condense.ensembl$gene_id)],
       pch=16, cex=0.5, col='deepskyblue')
points(df.compare$log2FC.MutDMSO.WtDMSO[which(df.compare$ensembl %in% mgi_condense.ensembl)],
       df.compare$log2FC.MutGSKJ4.MutDMSO[which(df.compare$ensembl %in% mgi_condense.ensembl)],
       pch=16, cex=0.5, col='magenta')
axis(1, seq(-4,4, by=2), tck=-0.02, cex.axis=0.7)
axis(2, seq(-4,4, by=2), tck=-0.02, cex.axis=0.7)
mtext('log2FC, R684C/+ DMSO vs +/+ DMSO', side=1, line=0.6, adj=0.5, cex=0.7)
mtext('log2FC, R684C/+ GSKJ4', side=2, line=1.2, adj=0.3, cex=0.7)
mtext('vs R684C/+ DMSO', side=2, line=0.6, adj=0.5, cex=0.7)
box(col='black')
abline(h=0, lty=2, col='red')
abline(v=0, lty=2, col='red')
abline(a=0, b=-1, col='blue')
dev.off()
```