---
title: "RNAseq +/- GSKJ4, MutGSKJ4 vs MutDMSO"
author: "Christine Gao"
date: "3/10/2023"
output: html_document
---

```{r, include=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(tximport)
library(DESeq2)
library(sva)
library(biomaRt)
library(ggplot2)
library(ggrepel)
library(tximeta)
library(SummarizedExperiment)
library(edgeR)
library(limma)
library(statmod)
library(tidyverse)
library(tripack)
```

Load dds object.
```{r}
load(file='dds_collapsed_GSKJ4.rda')
```

Set contrast for comparing MutGSKJ4 vs MutDMSO.
```{r}
mod_mat <- model.matrix(design(dds), colData(dds))
res.MutGSKJ4.MutDMSO <- results(dds, contrast = list(c('tx_GSKJ4_vs_DMSO',
                                      'genotypeMut.txGSKJ4')))
```

Plot p-value histogram.
```{r}
# setEPS()
# postscript('plots_final/pval-hist_MutGSKJ4.MutDMSO.eps')
hist(res.MutGSKJ4.MutDMSO$pvalue, 
     freq=FALSE,
     breaks = 40, 
     # ylim = c(0, 3), 
     # col = 'gray50', 
     xlab = "p-values", 
     main = "MutGSKJ4 vs MutDMSO", 
     font.main = 1,
     cex.axis=1.5,
     cex.lab=1.5)
# dev.off()
```

Get list of all differentially expresssed genes.
```{r}
diffexp.all <- as.data.frame(res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$padj < 0.1), ])

ensembl.id <- rownames(diffexp.all)

ensembl102.mmusculus <- useEnsembl(biomart='genes', dataset='mmusculus_gene_ensembl', version=102)
mgi <- getBM(attributes=c('ensembl_gene_id','mgi_symbol','mgi_description'),
             filters='ensembl_gene_id',
             values=ensembl.id,
             mart=ensembl102.mmusculus)
diffexp.all$ensembl_gene_id <- ensembl.id
diffexp.all <- merge(diffexp.all, mgi, by='ensembl_gene_id')
diffexp.all <- diffexp.all[order(diffexp.all$log2FoldChange, decreasing=TRUE),]

write.csv(diffexp.all, file='diffexp_MutGSKJ4-MutDMSO.csv')
```

Find corresponding unadjusted p-value for a threshold of FDR=0.1.
```{r}
fdr <- res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$padj>0.1 
                                  & is.na(res.MutGSKJ4.MutDMSO$padj)==FALSE),]
fdr[which(fdr$padj==min(fdr$padj)),]
```

Find max -log10(pvalue) to set ylim of plot.
```{r}
max(-log10(res.MutGSKJ4.MutDMSO$pvalue[which(is.na(res.MutGSKJ4.MutDMSO$pvalue) ==FALSE)]))
```

Create volcano plot function, plugging in the threshold p-value and the ylim.
```{r}
makeVolcanoPlot <- function(DEgenes_df, lfc_cutoff, gene_ids, circ_ids, circ_names, ylim, color){
  tab = data.frame(logFC = DEgenes_df$log2FoldChange, negLogPval = -log10(DEgenes_df$pvalue))
  rownames(tab) <- rownames(DEgenes_df)
  par(mar=c(5,5,5,5))
  plot(tab[-which(rownames(tab) %in% gene_ids), ], pch = 19, cex = 0.5, 
       xlab = expression(log[2]~fold~change),
       ylab = expression(-log[10]~pvalue), 
       cex.axis=1.5,
       cex.lab=1.5,
       col = "gray60", bty = 'l', 
       xlim = c(min(tab$logFC), max(tab$logFC)), 
       ylim = c(0, ylim))
  points(tab[gene_ids, ], pch = 19, cex = 0.5, col = color)
  points(tab[circ_ids, ], pch = 1, cex = 1.5, col= 'black')
  # circles(tab[circ_ids, ]$logFC, tab[circ_ids, ]$negLogPval, 
  #         r = rep(0.12, length(circ_ids)), col='black')
  text(tab[circ_ids, ]$logFC-0.15, tab[circ_ids, ]$negLogPval, 
       labels=circ_names, adj=1, cex = 0.75, col='black')
  abline(h = -log10(0.0222893), col= "red", lty=2, lwd=2)
  # legend <- legend('topleft', legend=c('FDR < 0.1','other'),
  #                  col=c(color,'gray60'), lty=c(NA, NA), pch=c(19,19), lwd=c(NA,NA), cex=1, bty='n')
  # legend <- legend('topright', legend=c('FDR = 0.1'),
  #                  col=c('cornflowerblue'), lty=c(2), pch=c(NA), lwd=c(2), cex=1.25, bty='n')
  if (lfc_cutoff != "none"){
    abline(v = c(-lfc_cutoff, lfc_cutoff), col = rgb(0,0,0,0.75), lty = "longdash", lwd = 1) 
  }
}
```

Make volcano plot, highlighting genes with FDR < 0.1. <i>\*removed points with padj=NA\*</i>
```{r}
# setEPS()
# postscript('plots_final/volcano_MutGSKJ4.MutDMSO.eps')
gene_ids <- rownames(res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$padj<0.1),])
makeVolcanoPlot(res.MutGSKJ4.MutDMSO[which(is.na(res.MutGSKJ4.MutDMSO$padj)==FALSE),], "none", gene_ids, circ_ids='', circ_names='', ylim=10, "orange")
# dev.off()
```

There is an extreme pvalue outlier, which is the point that led to a ylim of 32. Identify that point.
```{r}
res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$pvalue==min(res.MutGSKJ4.MutDMSO$pvalue)),]
```

Check counts for ENSMUSG00000024190, which is Dusp1, dual specificity phosphatase 1.
```{r}
counts(dds)['ENSMUSG00000024190', which(dds$genotype=='Mut' & dds$tx=='DMSO')]
counts(dds)['ENSMUSG00000024190', which(dds$genotype=='Mut' & dds$tx=='GSKJ4')]
```

There is also a log2FC outlier. Identify that point.
```{r}
res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$log2FoldChange==min(res.MutGSKJ4.MutDMSO$log2FoldChange)),]
```

Check counts for ENSMUSG00000092345, which is Gm20503, or predicted gene 20503.
```{r}
counts(dds)['ENSMUSG00000092345', which(dds$genotype=='Mut' & dds$tx=='DMSO')]
counts(dds)['ENSMUSG00000092345', which(dds$genotype=='Mut' & dds$tx=='GSKJ4')]
```

By ignoring the pvalue outliers, it becomes apparent that some points that meet FDR cutoff are not being color-marked appropriately.
```{r}
makeVolcanoPlot(res.MutGSKJ4.MutDMSO, "none", gene_ids, circ_ids='', circ_names='', ylim=10, "orange")
```

To find out which points and why:
```{r}
nrow(res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$padj < 0.1),])
nrow(res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$pvalue <= 0.0222893),])
```

There is discrepancy between the number of genes meeting the padj cutoff and those meeting the corresponding pvalue cutoff. Find out which genes are missing.
```{r}
data.frame(res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$pvalue < 0.0222893 &
                             !(rownames(res.MutGSKJ4.MutDMSO) %in% gene_ids)),])
```

The missing genes all have padj set to NA. Here it is explained that these genes were filtered by independent filtering for having a low mean normalized count (baseMean): https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA

and here:
https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#indfilt
```{r}
metadata(res.MutGSKJ4.MutDMSO)$alpha
metadata(res.MutGSKJ4.MutDMSO)$filterThreshold

plot(metadata(res.MutGSKJ4.MutDMSO)$filterNumRej, 
     type="b", ylab="number of rejections",
     xlab="quantiles of filter")
lines(metadata(res.MutGSKJ4.MutDMSO)$lo.fit, col="red")
abline(v=metadata(res.MutGSKJ4.MutDMSO)$filterTheta)
```

Make a volcano plot marking the Bmp pathway genes, in particular the 5 genes of interest. Almost all Bmp pathway genes are downregulated upon GSKJ4 treatment.
```{r}
load(file='gene_ids_bmp.rda')
gene_ids.bmp <- gene_ids
circ_ids <- c('ENSMUSG00000051279', 
              'ENSMUSG00000020427',
              'ENSMUSG00000021540',
              'ENSMUSG00000039153',
              'ENSMUSG00000060284')
circ_names <- c('Gdf6',
                'Igfbp3',
                'Smad5',
                'Runx2',
                'Sp7')
makeVolcanoPlot(res.MutGSKJ4.MutDMSO, "none", gene_ids, circ_ids, circ_names, ylim=10, "magenta")
```

Get the list of differentially expressed Bmp pathway genes.
```{r}
diffexp.bmp <- as.data.frame(res.MutGSKJ4.MutDMSO[which(res.MutGSKJ4.MutDMSO$padj < 0.1 &
                                            rownames(res.MutGSKJ4.MutDMSO) %in% gene_ids.bmp),])
ensembl.id <- rownames(diffexp.bmp)

ensembl102.mmusculus <- useEnsembl(biomart='genes', dataset='mmusculus_gene_ensembl', version=102)
mgi <- getBM(attributes=c('ensembl_gene_id','mgi_symbol','mgi_description'),
             filters='ensembl_gene_id',
             values=ensembl.id,
             mart=ensembl102.mmusculus)
diffexp.bmp$ensembl_gene_id <- ensembl.id
diffexp.bmp <- merge(diffexp.bmp, mgi, by='ensembl_gene_id')
diffexp.bmp <- diffexp.bmp[order(diffexp.bmp$log2FoldChange, decreasing=TRUE),]
write.csv(diffexp.bmp, file='diffexp_MutGSKJ4-MutDMSO_Bmp.csv')
```